annotell-input-api/annotell/input_api/cloud_storage/upload_handler.py

```diff
@@ -0,0 +1,69 @@
+import logging
+import time
+from pathlib import Path
+from typing import BinaryIO, Dict, Mapping, Optional
+
+import requests
+from requests.exceptions import HTTPError, ConnectionError
+from requests.models import Response
+
+from annotell.input_api.util import RETRYABLE_STATUS_CODES, get_content_type, get_wait_time
+
+log = logging.getLogger(__name__)
+
+
+class UploadHandler:
+
+    def __init__(self, max_retry_attempts: int = 23, max_retry_wait_time: int = 60, timeout: int = 60) -> None:
+        """
+        :param max_upload_retry_attempts: Max number of attempts to retry uploading a file to GCS.
+        :param max_upload_retry_wait_time:  Max with time before retrying an upload to GCS.
+        """
+        self.max_num_retries = max_retry_attempts
+        self.max_retry_wait_time = max_retry_wait_time  # seconds
+        self.timeout = timeout  # seconds
+
+    #  Using similar retry strategy as gsutil
+    #  https://cloud.google.com/storage/docs/gsutil/addlhelp/RetryHandlingStrategy
+    def _upload_file(self, upload_url: str, file: BinaryIO, headers: Dict[str, str], number_of_retries: int) -> None:
+        """
+        Upload the file to GCS, retries if the upload fails with some specific status codes.
+        """
+        log.info(f"Uploading file={file.name}")
+        resp = requests.put(upload_url, data=file, headers=headers, timeout=self.timeout)
+        try:
+            resp.raise_for_status()
+        except (HTTPError, ConnectionError) as e:
+            http_condition = number_of_retries > 0 and resp.status_code in RETRYABLE_STATUS_CODES
+            if http_condition or isinstance(e, ConnectionError):
+                self._handle_upload_error(upload_url, file, headers, resp, number_of_retries)
+            else:
+                raise e
+
+        except Exception as e:
+            raise e
+
+    def _handle_upload_error(self, upload_url: str, file: BinaryIO, headers: Dict[str, str], resp: Response, number_of_retries: int):
+        upload_attempt = self.max_num_retries - number_of_retries + 1
+        log.error(
+            f"On upload attempt ({upload_attempt}/{self.max_num_retries}) to GCS "
+            f"got response:\n{resp.status_code}: {resp.content}"
+        )
+        wait_time = get_wait_time(upload_attempt, self.max_retry_wait_time)
+        log.info(f"Waiting {int(wait_time)} seconds before retrying")
+        time.sleep(wait_time)
+        self._upload_file(upload_url, file, headers, number_of_retries - 1)
```

Comment:
```suggestion
        except (HTTPError, ConnectionError) as e:
            http_condition = number_of_retries > 0 and resp.status_code in RETRYABLE_STATUS_CODES
            if http_condition or isinstance(e, ConnectionError):
                self._handle_upload_error(resp, number_of_retries)
                self._upload_file(upload_url, file, headers, number_of_retries - 1)
            else:
                raise e

    def _handle_upload_error(self, resp: Response, number_of_retries: int):
        upload_attempt = self.max_num_retries - number_of_retries + 1
        log.error(
            f"On upload attempt ({upload_attempt}/{self.max_num_retries}) to GCS "
            f"got response:\n{resp.status_code}: {resp.content}"
        )
        wait_time = get_wait_time(upload_attempt, self.max_retry_wait_time)
        log.info(f"Waiting {int(wait_time)} seconds before retrying")
        time.sleep(wait_time)

```
I think this is neater. That way we don't have to pass all the parameters for uploading to the "waiting" method

Reply:
Also removed the unnecessary except handling

Reply:
I don't dare to commit this whole suggestion, but I'll be sure to implement it! ;D